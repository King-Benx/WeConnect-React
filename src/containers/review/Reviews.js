import React from 'react';
import { Button } from 'react-bootstrap';
import CustomFunctions from '../../custom/CustomFunctions';
import DashboardNavigation from '../../navigation/DashboardNavigation';
import { BASE_URL } from '../../custom/constants';
import axios from 'axios';
import BusinessInfo from '../business/BusinessInfo';
import ReviewInfo from './ReviewInfo';

class Reviews extends React.Component{
    constructor(props){
        super(props);
        this.state={
            review:"",
            reviews:[],
            business:[],
            id:this.props.match.params.id
        }
    }

    handleChange = (event) => {
        // handles change of state on input change
        this.setState({[event.target.id]: event.target.value});
    }

    componentDidMount(){
        // populates the business information and the reviews
        axios
        .get(BASE_URL+'api/v1/businesses/'+ this.state.id, 
        {headers:{ 'x-access-token':CustomFunctions.getToken()}})
        .then(res => {
            if (res.status === 404){
                this.setState({ business:[] });
            }else if(res.status === 200){
                this.setState({ business:res.data.business_info });
            }
        })
        .catch(err => {
            CustomFunctions.createNotifications(err.status, err.response.data.message);      
        });

        axios
        .get(BASE_URL+'api/v1/businesses/'+ this.state.id +'/reviews',
         {headers:{ 'x-access-token':CustomFunctions.getToken()}})
         .then(res => {
            if (res.status === 404){
                this.setState({ reviews:[] });
            }else if(res.status === 200){
                this.setState({ reviews:res.data.reviews.business_reviews });
            }
         })
        .catch(err => {
            CustomFunctions.createNotifications(err.status, err.response.data.message);     
        });
    };

    formSubmit = (event) => {
        // handles form submit of the reviews
        event.preventDefault();
        axios
            .post(BASE_URL+'api/v1/businesses/'+ this.state.id +'/reviews', { review:this.state.review },
            {headers:{ 'x-access-token':CustomFunctions.getToken()}})
            .then(res => {
                CustomFunctions.createNotifications(res.status, res.data.message);
                this.setState({ review:"" });
                this.componentDidMount();
            })
            .catch(err => {
                CustomFunctions.createNotifications(err.status, err.response.data.message);
            });
    }

     render(){
        const div_data= this.state.reviews.map(((review)=>{ 
            // iterates through the reviews object
            return(
                <ReviewInfo review={ review }/>
            );
        }));   
        return (
                <div className="row"> 
                    <DashboardNavigation/>
                    <div className="col-sm-9 content-wrapper">
                        <div className="row">
                            <div className="col-xs-12 col-sm-4">
                                <div className="row">
                                    <BusinessInfo business={ this.state.business } />
                                </div>
                            </div>
                            <div className="col-xs-12 col-sm-8">
                                {div_data}
                                <form method="POST" className="form" onSubmit={ this.formSubmit }>
                                    <legend className="text-info">Leave a review</legend>
                                    <div className="form-group">
                                        <div className="row">
                                            <div className="col-sm-9">
                                                <textarea name="review" id="review" value={ this.state.review } onChange={ this.handleChange } className="form-control" required="required" placeholder="Leave Review Here" ></textarea>
                                            </div>
                                            <div className="col-sm-3">
                                                <Button type="submit" bsStyle="success" block><i className="glyphicon glyphicon-plus" type="submit"></i> Add Review</Button>
                                            </div>
                                        </div>
                                    </div>
                                </form>
                            </div>      
                            </div>    
                        </div>
                    </div>
            );

    }
}
export default Reviews;