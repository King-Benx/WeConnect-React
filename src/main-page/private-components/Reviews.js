import React from 'react';
import {Redirect} from 'react-router-dom';
import { Button, Image } from 'react-bootstrap';
import CustomFunctions from '../functional-resources/CustomFunctions';
import DashboardNavigation from '../shared-components/DashboardNavigation';
import comment_icon from '../img/comment-icon.jpg';
import brief_case from '../img/briefcase.ico';
import { BASE_URL } from '../functional-resources/urls';
import superagent from 'superagent';

class Reviews extends React.Component{
    constructor(props){
        super(props);
        this.state={
            review:"",
            reviews:[],
            business:[],
            id:this.props.match.params.id
        }
    }

    handleChange = (event)=>{
        this.setState({[event.target.id]: event.target.value});
    }

    componentDidMount(){
        superagent
        .get(BASE_URL+'api/v1/businesses/'+this.state.id)
        .set({'x-access-token':JSON.parse(localStorage.getItem('data')).token})
        .end((err,res)=>{
            if(err){CustomFunctions.createNotifications(err.status, res.body.message)};
            if (res.status === 404){
                this.setState({business:[]});
            }else if(res.status === 200){
                this.setState({business:res.body.business_info});
            }
        });

        superagent
        .get(BASE_URL+'api/v1/businesses/'+this.state.id+'/reviews')
        .set({'x-access-token':JSON.parse(localStorage.getItem('data')).token})
        .end((err,res)=>{
            if(err){CustomFunctions.createNotifications(err.status, err.toString())};
            if (res.status === 404){
                this.setState({reviews:[]});
            }else if(res.status === 200){
                this.setState({reviews:res.body.reviews.business_reviews});
            }
        });
    };

    formSubmit(event){
        event.preventDefault();
        superagent
            .post(BASE_URL+'api/v1/businesses/'+this.state.id+'/reviews')
            .send({review:this.state.review})
            .set({'x-access-token':JSON.parse(localStorage.getItem('data')).token})
            .end((err,res)=>{
                if(err){CustomFunctions.createNotifications(err.status, err.toString());};
                CustomFunctions.createNotifications(res.status, res.body.message);
                this.setState();
                this.componentDidMount();
        });
    }

     render(){
        const div_data= this.state.reviews.map(((review)=>{
            return(
                <div className="row">
                    <div className="col-xs-2">
                        <Image src={comment_icon} alt="" responsive circle/>
                    </div>
                    <div className="col-xs-10">
                        <h4 className="list-group-item-heading">Author: <span className="text-primary">{review.author}</span></h4>
                        <blockquote className="list-group-item-text">{review.review}</blockquote>
                        <small className="text text-right text-primary">{review.last_modified}</small>
                        <hr/>
                    </div>
                    <br/>
                </div>
            );
        }));   
        return (
                CustomFunctions.checkAuthentication()? (   
                <div className="row"> 
                    <DashboardNavigation/>
                    <div className="col-sm-9 content-wrapper">
                        <div className="row">
                            <div className="col-xs-12 col-sm-4">
                                <div className="row">
                                    <div className="col-xs-12">
                                    <Image src={brief_case} alt="" thumbnail responsive/>
                                    <h4><b>Business name: </b> <span className="text text-info">{this.state.business.name}</span></h4>
                                    <h4><b>Category: </b> <span className="text text-info">{this.state.business.category}</span></h4>
                                    <h4><b>Location: </b> <span className="text text-info">{this.state.business.location}</span></h4>
                                    <h4><u>Description</u></h4>
                                    <p className="text text-info">{this.state.business.description}</p>
                                    </div>
                                </div>
                            </div>
                            <div className="col-xs-12 col-sm-8">
                                {div_data}
                                <form method="POST" className="form" onSubmit={this.formSubmit.bind(this)}>
                                    <legend className="text-info">Leave a review</legend>
                                    <div className="form-group">
                                        <div className="row">
                                            <div className="col-sm-9">
                                                <textarea name="review" id="review" value={this.state.review} onChange={this.handleChange.bind(this)} className="form-control" required="required" placeholder="Leave Review Here" ></textarea>
                                            </div>
                                            <div class="col-sm-3">
                                                <Button type="submit" bsStyle="success" block><i className="glyphicon glyphicon-plus"></i> Add Review</Button>
                                            </div>
                                        </div>
                                    </div>
                                </form>
                            </div>      
                            </div>    
                        </div>
                    </div>
                ): <Redirect to={{pathname:'/login'}}/>
            );

    }
}
export default Reviews;