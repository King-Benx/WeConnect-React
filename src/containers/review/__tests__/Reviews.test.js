import React from 'react';
import { shallow } from 'enzyme';
import Reviews from "../Reviews";
import renderer from 'react-test-renderer';

describe('These are tests for the review form ', () =>{
    let props;
    let mountedReviews;
    beforeEach(() => {
        props={
            match: {
                params: {
                    id:1
                }
            },
            business:{
                id:1,
                location:'location 1',
                description:'business description 1',
                category:'category 1',
                name:'business 1',
                user_id:1
            },
            review:{
                id:1,
                author:'test author',
                review:'test review',
                last_modified: '2018-01-01',
                business_id:'1',
            },
            id:'1'
        };
        mountedReviews = shallow(<Reviews { ...props } />)
    });
    const event={
        target:{
            value:{}
        },
        preventDefault: () => {

        }
    }

    it('renders without crashing', () => {
       shallow(<Reviews {...props} />)
       const tree = renderer.create(<Reviews {...props} />).toJSON();
       expect(tree).toMatchSnapshot();
    });

    it('handles input changed', () => {
        mountedReviews.instance().handleChange(event)
    })

    // it('handles form submit', () => {
    //     mountedReviews.instance().formSubmit(event)
    // })

    it('component did mount', () => {
        mountedReviews.instance().componentDidMount()
    })

    it('has a form',() =>{
        const forms = mountedReviews.find('form')
        expect(forms.length).toBe(1)
    });

    it('has a dashboard',() =>{
        const dashboards = mountedReviews.find('DashboardNavigation')
        expect(dashboards.length).toBe(1)
    });

    it('has business info',() =>{
        const business_info = mountedReviews.find('BusinessInfo')
        expect(business_info.length).toBe(1)
    });

    it('has buttons',() =>{
        const buttons = mountedReviews.find('Button')
        expect(buttons.length).toBe(1)
    });

    it('has a textarea',() =>{
        const textareas = mountedReviews.find('textarea')
        expect(textareas.length).toBe(1)
    });

    it('has divs',() =>{
        const divs = mountedReviews.find('div')
        expect(divs.length).toBe(10)
    });

    it('has a legend',() =>{
        const legends = mountedReviews.find('legend')
        expect(legends.length).toBe(1)
    });

});