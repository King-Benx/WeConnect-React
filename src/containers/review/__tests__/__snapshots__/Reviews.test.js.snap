// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`These are tests for the review form  renders without crashing 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-9 content-wrapper"
  >
    <div
      className="row"
    >
      <div
        className="col-xs-12 col-sm-4"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <img
              alt=""
              className="img-responsive img-thumbnail"
              src="briefcase.ico"
            />
            <h4>
              <b>
                Business name: 
              </b>
               
              <span
                className="text text-info"
              />
            </h4>
            <h4>
              <b>
                Category: 
              </b>
               
              <span
                className="text text-info"
              />
            </h4>
            <h4>
              <b>
                Location: 
              </b>
               
              <span
                className="text text-info"
              />
            </h4>
            <h4>
              <u>
                <b>
                  Description
                </b>
              </u>
            </h4>
            <p
              className="text text-info"
            />
          </div>
        </div>
      </div>
      <div
        className="col-xs-12 col-sm-8"
      >
        <form
          className="form"
          method="POST"
          onSubmit={[Function]}
        >
          <legend
            className="text-info"
          >
            Leave a review
          </legend>
          <div
            className="form-group"
          >
            <div
              className="row"
            >
              <div
                className="col-sm-9"
              >
                <textarea
                  className="form-control"
                  id="review"
                  name="review"
                  onChange={[Function]}
                  placeholder="Leave Review Here"
                  required="required"
                  value=""
                />
              </div>
              <div
                className="col-sm-3"
              >
                <button
                  className="btn btn-success btn-block"
                  disabled={false}
                  type="submit"
                >
                  <i
                    className="glyphicon glyphicon-plus"
                    type="submit"
                  />
                   Add Review
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
